### ISSUE TITLE
some pytroch mistake  in constraints.py about  " Adjusts features that are computed from other features." 

### ISSUE THREAD

**Issue by bigpigboomboomboom** [Mobile-Intelligence-Lab/Amoeba#1]
1

**Comment by bigpigboomboomboom** [comment]
> in constraints.py as follow code:
> 
> x_adv.data[:, 4] = x_adv.data[:, 1] * x_adv.data[:, 38] # feature 5 = 2 * 39 x_adv.data[:, 5] = x_adv.data[:, 2] * x_adv.data[:, 70] # feature 6 = 3 * 71 x_adv.data[:, 3] = x_adv.data[:, 4] + x_adv.data[:, 5] # feature 4 = 5 + 6 x_adv.data[:, 6] = x_adv.data[:, 3] / x_adv.data[:, 0] # feature 7 = 4 / 1
> 
> it does not actually change the value of x_adv[:,4],x_adv.data[:, 5],x_adv.data[:, 3], x_adv.data[:, 6] because manipulating tensor.data directly is not recommended. In PyTorch, slicing a tensor and directly assigning it to another slice does not modify the value of the original tensor. This is because x[:, 3].data returns a view of the tensor data, not the tensor itself
> 
> for example :
> 
> ```
> x = torch.tensor([[1,2,3],[4,5,6],[7,8,9]])
> x[:,0].data = x[:,1].data + x[:,2].data
> print(x)
> ```
> 
> result: tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
> 
> so i think it may change as follow : x_adv[:, 4] = (x_adv[:, 1] * x_adv[:, 38]).datech() # feature 5 = 2 * 39 x_adv[:, 5] = (x_adv[:, 2] * x_adv[:, 70]).datech() # feature 6 = 3 * 71 x_adv[:, 3] = (x_adv[:, 4] + x_adv[:, 5]).datech() # feature 4 = 5 + 6 x_adv[:, 6] = (x_adv[:, 3] / x_adv[:, 0]).datech() # feature 7 = 4 / 1
> 
> Features 4,5,3,6 don't actually need to keep track of the gradient, their values are calculated from other features , so their gradient can always be 0
